tools/hsmtool.c

```diff
@@ -368,6 +371,145 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
 	return 1;
 }
 
+static void get_words(struct words **words) {
+	printf("Select your language:\n");
+	printf("  0) English (en)\n");
+	printf("  1) Spanish (es)\n");
+	printf("  2) French (fr)\n");
+	printf("  3) Italian (it)\n");
+	printf("  4) Japanese (jp)\n");
+	printf("  5) Chinese simplified (zhs)\n");
+	printf("  6) Chinese traditional (zht)\n");
+	printf("Select [0-6]: ");
+
+	char *selection=NULL;
+	size_t size = 0;
+	size_t characters = getline(&selection, &size, stdin);
+	if (characters < 0)
+		errx(ERROR_USAGE, "Could not read line from stdin.");
+	int selected = atoi(selection);
+
+	char *lang = malloc(4);
+	switch (selected) {
+		case 0:
+			strcpy(lang, "en");
+			break;
+		case 1:
+			strcpy(lang, "es");
+			break;
+		case 2:
+			strcpy(lang, "fr");
+			break;
+		case 3:
+			strcpy(lang, "it");
+			break;
+		case 4:
+			strcpy(lang, "jp");
+			break;
+		case 5:
+			strcpy(lang, "zhs");
+			break;
+		case 6:
+			strcpy(lang, "zht");
+			break;
+		default:
+			errx(ERROR_USAGE, "Invalid language selection, select one from the list [0-6].");
+	}
+
+	bip39_get_wordlist(lang, words);
+}
+
+static void get_mnemonic(char *mnemonic) {
+	char *line = NULL;
+	size_t line_size = 0;
+
+	printf("Introduce your BIP39 word list:\n");
+	size_t characters = getline(&line, &line_size, stdin);
+	if (characters < 0)
+		errx(ERROR_USAGE, "Could not read line from stdin.");
+	line[characters-1] = '\0';
+	strcpy(mnemonic, line);
+	free(line);
+}
+
+static void read_mnemonic(char *mnemonic) {
+	/* Get words for the mnemonic language */
+	struct words *words;
+	get_words(&words);
+
+	/* Get mnemonic */
+	get_mnemonic(mnemonic);
+
+	if (bip39_mnemonic_validate(words, mnemonic) != 0) {
+		errx(ERROR_USAGE, "Invalid mnemonic: \"%s\"", mnemonic);
+	}
+}
+
+static void read_passphrase(char **passphrase) {
+	struct termios current_term, temp_term;
+	printf("Warning: remember that different passphrases yield different "
+	       "bitcoin wallets.\n");
+	printf("Leave it empty for using the value by default (echo is "
+	       "disabled now).\n");
+	printf("Enter your passphrase: \n");
+
+	/* Change terminal options so we do not echo the passphrase */
+	if (tcgetattr(fileno(stdin), &current_term) != 0)
+		errx(ERROR_USAGE, "Could not get current terminal options.");
+	temp_term = current_term;
+	temp_term.c_lflag &= ~ECHO;
+	if (tcsetattr(fileno(stdin), TCSAFLUSH, &temp_term) != 0)
+		errx(ERROR_USAGE, "Could not disable passphrase echoing.");
+	/* If we don't flush we might end up being buffered and we might seem
+	 * to hang while we wait for the password. */
+	fflush(stdout);
+
+	size_t passphrase_size = 0;
+	size_t characters = getline(passphrase, &passphrase_size, stdin);
+	if (characters < 0)
+		errx(ERROR_USAGE, "Could not read passphrase from stdin.");
+
+	/* Replace the last character because we assume it was the newline '\n' */
+	(*passphrase)[characters-1] = '\0';
```

Comment:
> Why do we assume it was ? Couldn't we just check it is ?

well, the user needs to introduce a value an press enter (`getline()`). I can add a check and fail in case something weird is going on.

> Also we should support `passphrase == NULL`, for the very common "no passphrase" case !

The empty word and NULL have the [same effect](https://github.com/ElementsProject/libwally-core/blob/c9c45b68c5a18ca1b5bca41f3c3c8350a23b1bd8/src/bip39.c#L209) in libwally. I doubt that this never changes but I can substitute it by NULL when the user does not add any character to their passphrase.

 > Also, characters can in theory be 0

How do we get the line then? ctrl-d or something? In that case I think it's ok to just `segfault`

Reply:
> well, the user needs to introduce a value an press enter (getline()). I can add a check and fail in case something weird is going on.

https://github.com/ElementsProject/lightning/pull/4065#discussion_r493036721 ... I know it's harder to get there with the two consecutive `getline`, but that's a tool and it's going to be used in scripts and stuff.
Moreover, you are overriding the user passphrase here! This definitely needs a sanity check (maybe a belt and suspenders, but at least the belt will do for now!).
If you are really confident about this, make it an assert:
```C
assert((*passphrase)[characters-1] == '\n');
(*passphrase)[characters-1] = '\0';
```
But i think it's a bad idea and can be triggered. Don't know what's wrong with just doing:
```C
if ((*passphrase)[characters-1] == '\n')
    (*passphrase)[characters-1] = '\0';
```

> The empty word and NULL have the same effect in libwally. I doubt that this never changes but I can substitute it by NULL when the user does not add any character to their passphrase.

Oh, i was not aware of that :) I just check the doc comment stating "if no passphrase then set to NULL". So yeah better to rely on documented behaviour i guess!

Reply:
Check added